File: CXOptimizeCollector/CodedUICollector/CXOptimizeCollector/config/ConfigurationLoader.cs
/*
 * Copyright 2014 - 2018 Cognizant Technology Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using com.cognizant.pace.CXOptimize.Collector.constant;
using com.cognizant.pace.CXOptimize.Collector.service;
using com.cognizant.pace.CXOptimize.Collector.utils;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace com.cognizant.pace.CXOptimize.Collector.config
{
    public class ConfigurationLoader
    {
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

        private static ConfigurationLoader instance = null;
        private static Object mutex = new Object();
        public Dictionary<String, Object> clientConfig = null;

        
        string propertyDirPath = string.Empty;

        private void GetPropertyFile(string fullPathName)
        {
            string lpropertyFilePath = fullPathName + "//collectordependency";

            if (!Directory.Exists(lpropertyFilePath))
            {
                GetPropertyFile(new DirectoryInfo(fullPathName).Parent.FullName);
            }
            else
            {
                propertyDirPath = lpropertyFilePath;
            }
                
        }
        /*
        private Dictionary<String, String> GetBeaconPropertyLoadConfig(String path)
        {

            string fullPathName = string.Empty;
            if (path != null)
            {
                fullPathName = propertyDirPath + "\\Collector.properties";
            }
            else
            {
                string projectDir = AppDomain.CurrentDomain.SetupInformation.ApplicationBase;
                LoggingMessage.DisplayMessage("CXOptimize - Project Target Dir Path -> " + projectDir, LoggingMessage.MessageType.DEBUG, log);
                GetPropertyFile(projectDir);

            }
            string[] split;
            string line = String.Empty;
            Dictionary<String, String> config = null;

            //string projectDir = AppDomain.CurrentDomain.SetupInformation.ApplicationBase;

          
            string environmentKey = ConfigurationManager.AppSettings["Environment"];
            if (environmentKey == null || environmentKey == string.Empty)
                fullPathName = propertyDirPath + "\\Collector.properties";
            else
                fullPathName = string.Format("{0}\\{1}_Collector.properties", propertyDirPath, environmentKey);

            LoggingMessage.DisplayMessage("CXOptimize - Config Path-> " + fullPathName, LoggingMessage.MessageType.DEBUG, log);

            if (File.Exists(fullPathName))
            {
                config = new Dictionary<string, string>();
                using (StreamReader file = new StreamReader(fullPathName))
                {
                    while ((line = file.ReadLine()) != null)
                    {
                        split = line.Split('=');
                        if (split.Length == 2)
                            config.Add(split[0], split[1]);
                    }
                }

            }
            else
            { 
                throw new ArgumentException("Collector Property file missing");
            }

            return config;
        }
        */

        protected ConfigurationLoader()
        {

            try
            {
                log.Debug("Starting Loading Configuration");

                String path = CollectorConstants.getCollectorProperties();

                if (path == null)
                {
                    GetPropertyFile(AppDomain.CurrentDomain.SetupInformation.ApplicationBase);
                    path = propertyDirPath + "\\Collector.properties";
                }
                else
                {
                    path = path + "\\Collector.properties";
                }

                Dictionary<String, String> config;
                string[] split;
                string line = String.Empty;

                if (File.Exists(path))
                {
                    config = new Dictionary<string, string>();
                    using (StreamReader file = new StreamReader(path))
                    {
                        while ((line = file.ReadLine()) != null)
                        {
                            split = line.Split('=');
                            if (split.Length == 2)
                                config.Add(split[0], split[1]);
                        }
                    }

                }
                else
                {
                    throw new ArgumentException("Collector Property file missing");
                }


                if (config != null)
                {
                    log.Debug("Properties Load Successfull from file : " + config.ToString());
                    CollectorConstants.setBeaconURL(config["beaconUrl"]);
                    CollectorConstants.setClientName(config["clientName"]);
                    CollectorConstants.setProjectName(config["projectName"]);
                    CollectorConstants.setScenarioName(config["scenarioName"]);
                    CollectorConstants.setLicenseKey(config["LicenseKey"]);
                    CollectorConstants.setUserName(config["UserName"]);
                    CollectorConstants.setPassword(config["Password"]);
                    /*
                    if (config.ContainsKey("Release") && config["Release"] != String.Empty)
                    {
                        CollectorConstants.setRelease(config["Release"].ToString());
                    }
                    else
                    {
                        CollectorConstants.setRelease("NA");
                    }
                    if (config.ContainsKey("Build") && config["Build"] != String.Empty)
                    {
                        CollectorConstants.setBuild(config["Build"].ToString());
                    }
                    else
                    {
                        CollectorConstants.setBuild("NA");
                    }*/
                    if (config.ContainsKey("isLoadTest") && config["isLoadTest"].Contains("true"))
                    {
                        CollectorConstants.setLoadTest("true");
                    }
                    else
                    {
                        CollectorConstants.setLoadTest("false");
                    }
                    if (config.ContainsKey("markWaitTime") && config["markWaitTime"] != String.Empty)
                    {
                        CollectorConstants.setMarkWaitTime(Int32.Parse(config["markWaitTime"]));
                    }
                    else
                    {
                        CollectorConstants.setMarkWaitTime(5000);
                    }
                    if (config.ContainsKey("resourceSettleTime"))
                    {
                        CollectorConstants.setResourceSettleTime(Int32.Parse(config["resourceSettleTime"]));
                    }
                    else
                    {
                        CollectorConstants.setResourceSettleTime(2000);
                    }
                    log.Debug("Loaded initialization parameters");
                }


                if (CollectorConstants.getBeaconURL() == null || CollectorConstants.getBeaconURL() == "")
                {
                    throw new ArgumentException("Missing value for BeaconURL");
                }
                if (CollectorConstants.getClientName() == null || CollectorConstants.getClientName() == "")
                {
                    throw new ArgumentException("Missing value for ClientName");
                }
                if (CollectorConstants.getProjectName() == null || CollectorConstants.getProjectName() == "")
                {
                    throw new ArgumentException("Missing value for ProjectName");
                }
                if (CollectorConstants.getScenarioName() == null || CollectorConstants.getScenarioName() == "")
                {
                    throw new ArgumentException("Missing value for ScenarioName");
                }
                if (CollectorConstants.getUserName() == null || CollectorConstants.getUserName() == "")
                {
                    throw new ArgumentException("Missing value for UserName");
                }
                if (CollectorConstants.getPassword() == null || CollectorConstants.getPassword() == "")
                {
                    throw new ArgumentException("Missing value for Password");
                }

                if (CollectorConstants.getMarkWaitTime() == 0)
                {
                    CollectorConstants.setMarkWaitTime(5000);
                }
                if (CollectorConstants.getResourceSettleTime() == 0)
                {
                    CollectorConstants.setResourceSettleTime(2000);
                }
                if (CollectorConstants.getRelease() == null || CollectorConstants.getRelease() == string.Empty)
                {
                    CollectorConstants.setRelease(DateTime.Now.ToString("yyyyMM"));
                }
                if (CollectorConstants.getBuild() == null || CollectorConstants.getBuild() == string.Empty)
                {
                    CollectorConstants.setBuild(DateTime.Now.ToString("yyyyMMddHH"));
                }


                Dictionary<string, object> configs = new Dictionary<string, object>();
                ICXOptimizeService cxOpService = new CXOptimizeServiceImpl();
                CollectorConstants.setApiToken(cxOpService.getAuthToken());

                log.Debug("Created ApiToken for the session " + CollectorConstants.getApiToken());

                configs = cxOpService.getConfiguration();

                if (config.ContainsKey("Status")) // Failed to get the config
                {
                    log.Info("Cannot load configuration: " + config["Reason"]);
                    config.Clear();
                }
                else
                {
                    configs.Add("beaconUrl", CollectorConstants.getBeaconURL());
                    configs.Add("RunTime", configs.ContainsKey("runTimestamp") ? configs["runTimestamp"] :
                    (DateTime.UtcNow.Subtract(new DateTime(1970, 1, 1))).TotalMilliseconds);
                    CollectorConstants.setStaticExt(configs.ContainsKey("staticResourceExtension") ? configs["staticResourceExtension"].ToString() : "");
                    CollectorConstants.setImages(configs.ContainsKey("imageResourceExtension") ? configs["imageResourceExtension"].ToString() : "");
                    CollectorConstants.setResDurThreshold((configs.ContainsKey("resourceDurationThreshold") ? Double.Parse(configs["resourceDurationThreshold"].ToString())
                        : Double.Parse("10")));
                    configs.Add("Release", CollectorConstants.getRelease());
                    configs.Add("BuildNumber", CollectorConstants.getBuild());
                    clientConfig = configs;

                    log.Debug("Complete Configuration " + config.ToString());
                }
            }
            catch (Exception e)
            {
                log.Error("Cannot load configuration due to exception : " + e);
            }
        }


        public static ConfigurationLoader getInstance()
        {
            if (instance == null)
            {
                lock (mutex)
                {
                    if (instance == null)
                    {
                        instance = new ConfigurationLoader();
                    }
                }
            }
            return instance;
        }
    }
}

File: CXOptimizeCollector/SeleniumCollector/Java/src/main/java/com/cognizant/pace/CXOptimize/Collector/config/ConfigurationLoader.java
/*
 * Copyright 2014 - 2018 Cognizant Technology Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.cognizant.pace.CXOptimize.Collector.config;

import com.cognizant.pace.CXOptimize.Collector.service.CXOptimizeService;
import com.cognizant.pace.CXOptimize.Collector.service.CXOptimizeServiceImpl;
import com.cognizant.pace.CXOptimize.Collector.constant.CollectorConstants;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.security.InvalidParameterException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Map;
import java.util.Properties;


public class ConfigurationLoader {
    private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(ConfigurationLoader.class);
    private static final Object mutex = new Object();
    private static ConfigurationLoader instance = null;
    public Map<String, Object> clientConfig = null;


    protected ConfigurationLoader() {

        try {
            LOGGER.debug("CXOP - Starting Loading Configuration");
            Properties mainProperties;
            mainProperties = new Properties();
            FileInputStream file;
            Map<String, Object> config;
            String licenseKey;
            String beaconURL;
            String authToken;
            String path = CollectorConstants.getCollectorProperties();

            //To integrate with Cognizant Intelligent Scripter
            if(path == null)
            {
                path = System.getProperty("user.dir");
                path = path + File.separator + "collectordependency" + File.separator + "Collector.properties";
            }
            else
            {
                path =  path + File.separator + "Collector.properties";
            }

            LOGGER.debug("CXOP - Checking if the properties file exists in {}", path);
            File propFile = new File(path);
            boolean exists = propFile.exists();
            LOGGER.debug("CXOP - Properties file exists : {}", exists);

            if (exists) {
                file = new FileInputStream(path);
                mainProperties.load(file);
                LOGGER.debug("CXOP - Properties Load Successfull from file : {}", mainProperties.toString());
                CollectorConstants.setBeaconURL(mainProperties.getProperty("beaconUrl"));
                CollectorConstants.setClientName(mainProperties.getProperty("clientName"));
                CollectorConstants.setProjectName(mainProperties.getProperty("projectName"));
                CollectorConstants.setScenarioName(mainProperties.getProperty("scenarioName"));
                CollectorConstants.setLicenseKey(mainProperties.getProperty("LicenseKey"));
                CollectorConstants.setUserName(mainProperties.getProperty("UserName"));
                CollectorConstants.setPassword(mainProperties.getProperty("Password"));

                if (mainProperties.containsKey("isLoadTest") && mainProperties.getProperty("isLoadTest").equals("true")) {
                    CollectorConstants.setLoadTest("true");
                }
                else
                {
                    CollectorConstants.setLoadTest("false");
                }
                if (mainProperties.containsKey("markWaitTime")) {
                    CollectorConstants.setMarkWaitTime(Integer.parseInt(mainProperties.getProperty("markWaitTime").toString()));
                } else {
                    CollectorConstants.setMarkWaitTime(5000);
                }
                if (mainProperties.containsKey("resourceSettleTime")) {
                    CollectorConstants.setResourceSettleTime(Integer.parseInt(mainProperties.getProperty("resourceSettleTime").toString()));
                } else {
                    CollectorConstants.setResourceSettleTime(2000);
                }
                if (mainProperties.containsKey("manualResourceTimeClear")) {
                    CollectorConstants.setManualResourceTimeClear(mainProperties.getProperty("manualResourceTimeClear"));
                } else {
                    CollectorConstants.setManualResourceTimeClear("false");
                }
                LOGGER.debug("CXOP - Loaded initialization parameters");
                file.close();
            } else {
                mainProperties.load(ConfigurationLoader.class.getResourceAsStream("/collectordependency/Collector.properties"));
                if (mainProperties.getProperty("clientName") != null) {
                    LOGGER.debug("CXOP - Properties Load Successfull from package : {}", mainProperties.toString());
                    CollectorConstants.setBeaconURL(mainProperties.getProperty("beaconUrl"));
                    CollectorConstants.setClientName(mainProperties.getProperty("clientName"));
                    CollectorConstants.setProjectName(mainProperties.getProperty("projectName"));
                    CollectorConstants.setScenarioName(mainProperties.getProperty("scenarioName"));
                    CollectorConstants.setLicenseKey(mainProperties.getProperty("LicenseKey"));
                    CollectorConstants.setUserName(mainProperties.getProperty("UserName"));
                    CollectorConstants.setPassword(mainProperties.getProperty("Password"));

                    if (mainProperties.containsKey("isLoadTest") && mainProperties.getProperty("isLoadTest").equals("true")) {
                        CollectorConstants.setLoadTest("true");
                    }
                    else
                    {
                        CollectorConstants.setLoadTest("false");
                    }
                    if (mainProperties.containsKey("markWaitTime")) {
                        CollectorConstants.setMarkWaitTime(Integer.parseInt(mainProperties.getProperty("markWaitTime").toString()));
                    } else {
                        CollectorConstants.setMarkWaitTime(5000);
                    }

                    if (mainProperties.containsKey("resourceSettleTime")) {
                        CollectorConstants.setResourceSettleTime(Integer.parseInt(mainProperties.getProperty("resourceSettleTime").toString()));
                    } else {
                        CollectorConstants.setResourceSettleTime(2000);
                    }

                    if (mainProperties.containsKey("manualResourceTimeClear")) {
                        CollectorConstants.setManualResourceTimeClear(mainProperties.getProperty("manualResourceTimeClear"));
                    } else {
                        CollectorConstants.setManualResourceTimeClear("false");
                    }

                    LOGGER.debug("CXOP - Loaded initialization parameters");
                } else {
                    LOGGER.debug("CXOP - Properties loaded from environment variables");
                }
            }

            if (CollectorConstants.getBeaconURL() == null || CollectorConstants.getBeaconURL() == "") {
                throw new InvalidParameterException("Missing value for BeaconURL");
            }
            if (CollectorConstants.getClientName() == null || CollectorConstants.getClientName() == "") {
                throw new InvalidParameterException("Missing value for ClientName");
            }
            if (CollectorConstants.getProjectName() == null || CollectorConstants.getProjectName() == "") {
                throw new InvalidParameterException("Missing value for ProjectName");
            }
            if (CollectorConstants.getScenarioName() == null || CollectorConstants.getScenarioName() == "") {
                throw new InvalidParameterException("Missing value for ScenarioName");
            }
            if (CollectorConstants.getUserName() == null || CollectorConstants.getUserName() == "") {
                throw new InvalidParameterException("Missing value for UserName");
            }
            if (CollectorConstants.getPassword() == null || CollectorConstants.getPassword() == "") {
                throw new InvalidParameterException("Missing value for Password");
            }
            if (CollectorConstants.getMarkWaitTime() == 0) {
                CollectorConstants.setMarkWaitTime(5000);
            }
            if (CollectorConstants.getResourceSettleTime() == 0) {
                CollectorConstants.setResourceSettleTime(2000);
            }

            if (CollectorConstants.getManualResourceTimeClear().equals("true")) {
                CollectorConstants.setManualResourceTimeClear("true");
            }
            else
            {
                CollectorConstants.setManualResourceTimeClear("false");
            }
            if(CollectorConstants.getRelease() == null || CollectorConstants.getRelease() == "")
            {
                CollectorConstants.setRelease(new SimpleDateFormat("yyyyMM").format(new Date()));
            }
            if(CollectorConstants.getBuild() == null || CollectorConstants.getBuild() == "")
            {
                CollectorConstants.setBuild(new SimpleDateFormat("yyyyMMddHH").format(new Date()));
            }




            CXOptimizeService cxOpService = new CXOptimizeServiceImpl();
            CollectorConstants.setApiToken(cxOpService.getAuthToken());

            LOGGER.debug("CXOP - Created ApiToken for the session {}", CollectorConstants.getApiToken());

            config = cxOpService.getConfiguration();

            config.put("beaconUrl", CollectorConstants.getBeaconURL());
            config.put("RunTime", config.containsKey("runTimestamp") ? config.get("runTimestamp") : System.currentTimeMillis());
            config.put("Release",CollectorConstants.getRelease());
            config.put("BuildNumber",CollectorConstants.getBuild());

            CollectorConstants.setStaticExt(config.get("staticResourceExtension").toString());
            CollectorConstants.setImages(config.get("imageResourceExtension").toString());
            CollectorConstants.setResDurThreshold((config.get("resourceDurationThreshold") != null ? Double.parseDouble(config.get("resourceDurationThreshold").toString()) : Double.parseDouble("10")));

            clientConfig = config;

            LOGGER.debug("CXOP - Complete Configuration :{}", config.toString());
        } catch (Exception e) {
            LOGGER.error("CXOP - Cannot load configuration due to exception : {}",e);
        }
    }

    @SuppressWarnings("DoubleCheckedLocking")
    public static ConfigurationLoader getInstance() {
        if (instance == null) {
            synchronized (mutex) {
                if (instance == null) {
                    instance = new ConfigurationLoader();
                }
            }
        }
        return instance;
    }

    //To integrate with Cognizant Intelligent Scripter
    public static void reloadConfig()
    {
        instance = null;
    }

}
